link/perf_event.go

```diff
@@ -165,61 +229,68 @@ func (pe *perfEvent) Close() error {
-	if prog == nil {
-		return errors.New("cannot attach a nil program")
-	}
-	if pe.fd == nil {
-		return errors.New("cannot attach to nil perf event")
-	}
-	if prog.FD() < 0 {
-		return fmt.Errorf("invalid program: %w", sys.ErrClosedFd)
-	}
-	switch pe.typ {
-	case kprobeEvent, kretprobeEvent, uprobeEvent, uretprobeEvent:
-		if t := prog.Type(); t != ebpf.Kprobe {
-			return fmt.Errorf("invalid program type (expected %s): %s", ebpf.Kprobe, t)
+func attachPerfEvent(l Link, prog *ebpf.Program) (Link, error) {
+	validate := func(pe *perfEvent, prog *ebpf.Program) error {
+		if prog == nil {
+			return errors.New("cannot attach a nil program")
 		}
-	case tracepointEvent:
-		if t := prog.Type(); t != ebpf.TracePoint {
-			return fmt.Errorf("invalid program type (expected %s): %s", ebpf.TracePoint, t)
+		if prog.FD() < 0 {
+			return fmt.Errorf("invalid program: %w", sys.ErrClosedFd)
 		}
-	default:
-		return fmt.Errorf("unknown perf event type: %d", pe.typ)
+		if pe.fd == nil {
+			return errors.New("cannot attach to nil perf event")
+		}
+		switch pe.typ {
+		case kprobeEvent, kretprobeEvent, uprobeEvent, uretprobeEvent:
+			if t := prog.Type(); t != ebpf.Kprobe {
+				return fmt.Errorf("invalid program type (expected %s): %s", ebpf.Kprobe, t)
+			}
+		case tracepointEvent:
+			if t := prog.Type(); t != ebpf.TracePoint {
+				return fmt.Errorf("invalid program type (expected %s): %s", ebpf.TracePoint, t)
+			}
+		default:
+			return fmt.Errorf("unknown perf event type: %d", pe.typ)
+		}
+		return nil
 	}
 
-	pfd := pe.fd
-
-	if err := haveBPFLinkPerfEvent(); err == nil {
-		// Use the bpf api to attach the perf event (BPF_LINK_TYPE_PERF_EVENT, 5.15+).
-		//
-		// https://github.com/torvalds/linux/commit/b89fbfbb854c9afc3047e8273cc3a694650b802e
-		attr := sys.LinkCreatePerfEventAttr{
-			ProgFd:     uint32(prog.FD()),
-			TargetFd:   pfd.Uint(),
-			AttachType: sys.BPF_PERF_EVENT,
-			BpfCookie:  pe.cookie,
-		}
+	pe, ok := l.(*perfEvent)
+	if !ok {
+		pl, _ := l.(*perfEventLink)
 
-		fd, err := sys.LinkCreatePerfEvent(&attr)
-		if err != nil {
-			return fmt.Errorf("cannot create bpf perf link: %v", err)
-		}
-		pe.bpfLinkFD = fd
-	} else {
-		if pe.cookie != 0 {
-			return fmt.Errorf("bpf cookies are not available: %w", ErrNotSupported)
+		if err := validate(pl.pe, prog); err != nil {
+			pl.pe.Close()
+			return nil, err
 		}
 
-		// Assign the eBPF program to the perf event.
-		err := unix.IoctlSetInt(pfd.Int(), unix.PERF_EVENT_IOC_SET_BPF, prog.FD())
-		if err != nil {
-			return fmt.Errorf("setting perf event bpf program: %w", err)
-		}
+		return pl, attachPerfEventLink(pl, prog)
+	}
+
+	if pe.cookie != 0 {
+		pe.Close()
+		return nil, fmt.Errorf("bpf cookies are not available: %w", ErrNotSupported)
+	}
+
+	if err := validate(pe, prog); err != nil {
+		pe.Close()
+		return nil, err
+	}
+
+	return pe, attachPerfEventIoctl(pe, prog)
+}
+
+func attachPerfEventIoctl(pe *perfEvent, prog *ebpf.Program) error {
```

Comment:
The signature should be `(pe *perfEvent, prog *ebpf.Program) (*perfEventIoctl, error)`.